src/api/procedures/addInstruction.ts

```diff
@@ -98,128 +118,198 @@ export async function prepareAddInstruction(
     });
   }
 
-  if (!legs.length) {
+  const addAndAffirmInstructions: [
+    u64,
+    SettlementType,
+    Moment | null,
+    Moment | null,
+    {
+      from: PortfolioId;
+      to: PortfolioId;
+      asset: Ticker;
+      amount: Balance;
+    }[],
+    PortfolioId[]
+  ][] = [];
+  const addInstructions: [
+    u64,
+    SettlementType,
+    Moment | null,
+    Moment | null,
+    {
+      from: PortfolioId;
+      to: PortfolioId;
+      asset: Ticker;
+      amount: Balance;
+    }[]
+  ][] = [];
+
+  const noLegs: number[] = [];
+  const noFutureBlock: number[] = [];
+  const noFutureDate: number[] = [];
+
+  await P.each(instructions, async ({ legs, endBlock, tradeDate, valueDate }, i) => {
+    if (!legs.length) {
+      noLegs.push(i + 1);
+    }
+
+    let endCondition;
+
+    if (endBlock) {
+      const latestBlock = await context.getLatestBlock();
+
+      if (endBlock.lte(latestBlock)) {
+        noFutureBlock.push(i + 1);
+      }
+
+      endCondition = { type: InstructionType.SettleOnBlock, value: endBlock } as const;
+    } else {
+      endCondition = { type: InstructionType.SettleOnAffirmation } as const;
+    }
+
+    if (tradeDate && valueDate && tradeDate > valueDate) {
+      noFutureDate.push(i + 1);
+    }
+
+    if (!noLegs.length && !noFutureBlock.length && !noFutureDate.length) {
+      const rawVenueId = numberToU64(venueId, context);
+      const rawSettlementType = endConditionToSettlementType(endCondition, context);
+      const rawTradeDate = tradeDate ? dateToMoment(tradeDate, context) : null;
+      const rawValueDate = valueDate ? dateToMoment(valueDate, context) : null;
```

Comment:
COMPLEXITY: if you default both tradeDate and valueDate to null, you can transform these into

`const rawTradeDate = tradeDate && dateToMoment(tradeDate, context)`